x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,212 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package awss3
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common/acker"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+)
+
+var ackerIDCounter *atomic.Uint64
+
+func init() {
+	ackerIDCounter = atomic.NewUint64(0)
+}
+
+// EventACKTracker tracks the publishing state of S3 objects. Specifically
+// it tracks the number of message acknowledgements that are pending from the
+// output. It can be used to wait until all ACKs have been received for one or
+// more S3 objects.
+type EventACKTracker struct {
+	ID uint64
+
+	EventsAcked       *atomic.Uint64
+	EventsDropped     *atomic.Uint64
+	EventsPublished   *atomic.Uint64
+	EventsToBeTracked *atomic.Uint64
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	deletionWg *sync.WaitGroup
+
+	mutex *sync.Mutex
+
+	msg             *types.Message
+	ReceiveCount    int
+	start           time.Time
+	processingErr   error
+	Handles         []s3ObjectHandler
+	keepaliveCancel context.CancelFunc
+	keepaliveWg     *sync.WaitGroup
+	msgHandler      sqsProcessor
+	log             *logp.Logger
+}
+
+func NewEventACKTracker(ctx context.Context, deletionWg *sync.WaitGroup) *EventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	acker := &EventACKTracker{
+		ID:                ackerIDCounter.Inc(),
+		ctx:               ctx,
+		cancel:            cancel,
+		deletionWg:        deletionWg,
+		mutex:             new(sync.Mutex),
+		EventsAcked:       atomic.NewUint64(0),
+		EventsDropped:     atomic.NewUint64(0),
+		EventsPublished:   atomic.NewUint64(0),
+		EventsToBeTracked: atomic.NewUint64(0),
+	}
+
+	go func() {
+		t := time.NewTicker(500 * time.Microsecond)
+		defer t.Stop()
+
+		for {
+			<-t.C
+			if !acker.FullyTracked() {
+				continue
+			}
+
+			acker.cancelAndFlush()
+			return
+		}
+	}()
+
+	return acker
+}
+
+func (a *EventACKTracker) cancelAndFlush() {
+	a.cancel()
+	a.FlushForSQS()
+}
+
+// MarkSQSProcessedWithData Every call after the first one is a no-op
+func (a *EventACKTracker) MarkSQSProcessedWithData(msg *types.Message, publishedEvent uint64, receiveCount int, start time.Time, processingErr error, handles []s3ObjectHandler, keepaliveCancel context.CancelFunc, keepaliveWg *sync.WaitGroup, msgHandler sqsProcessor, log *logp.Logger) {
+	// We want to execute the logic of this call only once, when the ack mutex was locked on init
+	if a.EventsToBeTracked.Load() > 0 {
+		return
+	}
+
+	a.msg = msg
+	a.EventsToBeTracked = atomic.NewUint64(publishedEvent)
```

Comment:
`EventsToBeTracked` is a pointer to an atomic, not an atomic, and reassigning a pointer is not guaranteed to be atomic. For atomic assignment, use `EventsToBeTracked.Store`.

Probably all of the atomics in `EventACKTracker` should be type `atomic.Uint64`, not `*atomic.Uint64`, since pointers can encourage this kind of mismatch.